core/shark_turbine/kernel/compiler/builder.py

```diff
@@ -217,10 +238,33 @@ def unary_vector_arithmetic(self, op: str, val: IRProxyValue) -> IRProxyValue:
 
     ### Specializations
 
-    def promote_index_to_f32(self, value: Value, to_type: IrType) -> Value:
-        i32_type = IntegerType.get_signless(32)
-        i32 = arith_d.index_cast(i32_type, value)
-        return arith_d.sitofp(to_type, i32)
+    # Casting
+    def to_dtype_index_to_integer(self, value: Value, to_type: IrType) -> Value:
+        return arith_d.index_cast(to_type, value)
+
+    def to_dtype_index_to_float(self, value: Value, to_type: IrType) -> Value:
+        # Cast index to integer, and then ask for a integer to float cast.
+        # TODO: I don't really know how to query the machine bitwidth here,
+        # so using 64.
+        casted_to_int = arith_d.index_cast(IntegerType.get_signless(64), value)
+        return self.to_dtype(IRProxyValue(casted_to_int), to_type).ir_value
+
+    def to_dtype_integer_to_float(self, value: Value, to_type: IrType) -> Value:
+        # sitofp
+        casted_to_float = arith_d.sitofp(to_type, value)
+        return self.to_dtype(IRProxyValue(casted_to_float), to_type).ir_value
+
+    def to_dtype_float_to_float(self, value: Value, to_type: IrType) -> Value:
+        # Check bitwidth to determine if we need to extend or narrow
+        from_type = value.type
+        from_bitwidth = self.get_float_bitwidth(from_type)
+        to_bitwidth = self.get_float_bitwidth(to_type)
+        if from_bitwidth < to_bitwidth:
+            return arith_d.extf(to_type, value)
+        elif from_bitwidth > to_bitwidth:
+            return arith_d.truncf(to_type, value)
+        else:
+            raise CodegenError(f"NYI: Cast from {from_type} to {to_type}")
```

Comment:
I don't know how to convert between bf16 and fp16

Reply:
No one does :) bf16 was made in protest of fp16...

Best that can be done is cast to f32 then to f16. The former is just bit twiddling.

Reply:
Or leave it unimplemented with error.

Reply:
welp, leaving it unimplemented then